### ISSUE TITLE
[Bug] AWS Connection.py Using Private Botocore `get_response` Which Expects JSON But is Being Passed XML

### ISSUE THREAD

**Issue by o-nikolas** [celery/kombu#1726]
Hello Kombu folks!

### tl;dr
Kombu is using an internal/private method, [get_response](https://github.com/boto/botocore/blob/develop/botocore/response.py#L180-L201), from Botocore which changed behaviour recently due to other Botocore updates to the wire protocol used for SQS, which breaks Kombu.
The change has been temporarily reverted, but Kombu cannot depend on a stable behaviour of this method going forward as the changes will soon be applied again.

### Full Context/Explanation
The implementation of the AWS asynchronous connection is currently using the non-public method `get_response` from `botocore`, here is an example:
https://github.com/celery/kombu/blob/3e098dc94ed2a389276ccf3606a0ded3da157d72/kombu/asynchronous/aws/connection.py#L232-L236

`get_response` is vendored in here:
https://github.com/celery/kombu/blob/3e098dc94ed2a389276ccf3606a0ded3da157d72/kombu/asynchronous/aws/ext.py#L1-L9

Botocore recently switched to JSON for the wire protocol it uses to communicate with SQS [here](https://github.com/boto/botocore/commit/3e1f77d8c8eae589cdfb4b9ec09c816570f19904#diff-53c9ed27fbb58a2b8274cc8203f6dbbf167840984a94be3b55df322762f01b5bL6-R15) (Note: Commit url doesn't expand nicely, you'll need to follow the link).

Kombu builds XML AWS Query API requests manually (instead of using something like the boto3 SDK client) to communicate with SQS:
https://github.com/celery/kombu/blob/3e098dc94ed2a389276ccf3606a0ded3da157d72/kombu/asynchronous/aws/connection.py#L185-L203

The resulting XML responses are then passed to the aforementioned `get_response`, which now expects JSON due to the change in Botocore also linked above, for parsing. **In this situation, `get_response` returns a response object with an empty body, which Kombu reads to be a legitimate empty result. This means that a request to get messages, for example, will always return empty. In this way it fails silently.** 

Botocore has reverted the change [here](https://github.com/boto/botocore/commit/47676c3310601dc2b1a1132799e9cc975d386b4a). But this is only a temporary fix, as JSON is the preferred wire protocol.

### Possible Paths Forward
1. Kombu could move to using the public boto3 SDK to interact with SQS (create queues, read messages, etc) instead of hand crafting AWS Query API requests.
   **pros**: Seamless support from a public interface which will handle the wire protocol migration smoothly
   **cons**: This is a large refactoring, and presumably, there is some historical reason that we're unaware of for why Kombu is hand crafting requests since it is a complex piece of code and there was likely a reason for the effort to build it this way.
1. Re-institute the XML parsing that was previously in place before the migration to `get_response` (which was done in [this PR here](https://github.com/celery/kombu/pull/693/files#diff-a33f9d4b5ed32656d1d451c1aed5a1cc2b73969ed6af85df50c2efed6de73cfdL262-R234)).
   **pros**: Smaller scope for the change which could likely be completed sooner. Code worked in the past
   **cons**: More complex handcrafted request/response handling existing in Kombu rather than simply using boto3 SDK
1. Something else entirely? Please let us know what you think!

Thanks in advance for your time :pray:   
 
CC:
- MWAA: @kytha, @rafidka, @dhegberg, @yumike2006
- Airflow: @onikolas, @vincbeck
- Boto: @nateprewitt

**Comment by vincbeck** [comment]
Nice :) We also think this is the "best" solution. I am not sure what you are asking by "release notes" though. You can find documentation of boto3 [here](https://boto3.amazonaws.com/v1/documentation/api/latest/index.html). Boto3 interacts will all (or at least most) AWS services, you can find the documentation specific to SQS [here](https://boto3.amazonaws.com/v1/documentation/api/latest/reference/services/sqs.html). The documentation is pretty well done, you have specification for all APIs with examples. Feel free to ping us if you need any help/guidance

**Comment by auvipy** [comment]
> Nice :) We also think this is the "best" solution. I am not sure what you are asking by "release notes" though. You can find documentation of boto3 [here](https://boto3.amazonaws.com/v1/documentation/api/latest/index.html). Boto3 interacts will all (or at least most) AWS services, you can find the documentation specific to SQS [here](https://boto3.amazonaws.com/v1/documentation/api/latest/reference/services/sqs.html). The documentation is pretty well done, you have specification for all APIs with examples. Feel free to ping us if you need any help/guidance

I didn't know that we are using non public API!

**Comment by auvipy** [comment]
can you aws expert guys have a look at this issue? https://github.com/celery/celery/issues/8015#issuecomment-1555962253

**Comment by vincbeck** [comment]
Happy to help on this issue indeed but it is not related to this issue right? At least, to me, it does not look related.

Have you, by any chance, started to work on the `get_response` issue? Or any plan to address it?

**Comment by o-nikolas** [comment]
> can you aws expert guys have a look at this issue? [celery/celery#8015 (comment)](https://github.com/celery/celery/issues/8015#issuecomment-1555962253)

The above issue is not the same as this issue, since the cause for this issue was rolled back. However, after a quick look, the above does seem like another issue related to hand crafting http requests instead of simply using the boto3 SDK

**Comment by shubham22** [comment]
@auvipy - I noticed that this issue is marked for the 5.3.x release, which is due by June 30th. I'm curious to know whether the Kombu community is planning to address this in the next release? If not, is there anything we can do to get this prioritized?

**Comment by auvipy** [comment]
we have lot on the plate already. this is a big overhaul. but I will try

**Comment by rafidka** [comment]
@auvipy , I am from the Amazon MWAA team. I experimented with this and managed to do a PoC by making `kombu` use `boto3` library instead of manually crafting a request (path no. 1 in @o-nikolas original message). I confirmed it solves the issue, so I am happy to publish a PR for this. However, there are a couple of issues that I want to discuss with you first.

Basically, my solution involves updating [the `AsyncSQSConnection` class](https://github.com/celery/kombu/blob/main/kombu/asynchronous/aws/sqs/connection.py#LL16C11-L16C11) to use `boto3`. One challenge, though, is that `boto3` library is not async (and I wonder whether that was a contributing factor initially for deciding to manually craft the AWS request). However, this can be easily addressed by employing `kombu`'s event loop. Below is my re-implementation of the [`receive_message` method](https://github.com/celery/kombu/blob/main/kombu/asynchronous/aws/sqs/connection.py#L59-L77) employing it:

```python
    def receive_message(
        self, queue_url, number_messages=1, visibility_timeout=None,
        attributes=('ApproximateReceiveCount',), wait_time_seconds=None,
        callback=None
    ):
        def make_request(
            queue_url, number_messages, wait_time_seconds, visibility_timeout,
            attributes, callback
        ):
            kwargs = {
                "QueueUrl": queue_url,
                "MaxNumberOfMessages": number_messages,
                "MessageAttributeNames": attributes,
                "WaitTimeSeconds": wait_time_seconds,
            }
            if visibility_timeout:
                kwargs["VisibilityTimeout"] = visibility_timeout
            resp = self.sqs_connection.receive_message(**kwargs)
            if callback:
                callback(resp)

        # Let kombu's event loop call `make_request` asynchronously.
        return self.hub.call_soon(
            make_request, 
            queue_url, number_messages, wait_time_seconds, visibility_timeout,
            attributes, callback,
        )
```

I tested this and confirmed it is fixing the issue. However, one thing that still needs to be tested is with regard to multi-threading and multi-processing. `kombu`'s SQS broker [uses a boto3 session](https://github.com/celery/kombu/blob/main/kombu/transport/SQS.py#L674-L679) to communicate with SQS. Unfortunately, boto3 sessions (and resources) are [not thread safe](https://boto3.amazonaws.com/v1/documentation/api/1.19.0/guide/clients.html#multithreading-or-multiprocessing-with-clients); only clients are. Furthermore, when it comes to multi-processing all of them are not safe to be shared.

Based on my testing, the same process and thread execute both the `call_soon` and `make_request` methods, which makes sense since the SQS broker [relies on the event loop](https://github.com/celery/kombu/blob/main/kombu/transport/SQS.py#L555) to communicate with SQS, and my modified version of `receive_message` also relies on the same event loop. However:

1. Is it possible that the same event loop might end up employing more than one thread or process for some reason? I quickly checked the implementation in [hub.py](https://github.com/celery/kombu/blob/main/kombu/asynchronous/hub.py) and that doesn't seem to be the case.

2. Is it possible to have more than one event loop? That doesn't seem to be the case either, since the [`set_event_loop` method](https://github.com/celery/kombu/blob/main/kombu/asynchronous/hub.py#L49-L53) uses a global variable, but I cannot be fully confident considering my limited understanding of Celery code base.

3. Even if the answer to both questions above is "No", I am not very comfortable about `SQS.py` making the assumption that the event loop is going to be unique, as this is an implicit assumption and, worse, it is very hard to discover without in depth understanding of how async IO is implemented within `kombu`. So, it is theoretically possible that in the future multiple event loops are introduced (assuming that is not the case already) to achieve more IO concurrency.

Looking forward to hearing from you soon. I am happy to discuss this with you over a meeting to move things faster.

**Comment by auvipy** [comment]
> @auvipy , I am from the Amazon MWAA team. I experimented with this and managed to do a PoC for a solution by making `kombu` use `boto3` library instead of manually crafting a request (path no. 1 in @o-nikolas original message). I confirmed it solves the issue, so I am happy to publish a PR for this. However, there are a couple of issues that I want to discuss with you first.
> 
> Basically, my solution involves updating [the `AsyncSQSConnection` class](https://github.com/celery/kombu/blob/main/kombu/asynchronous/aws/sqs/connection.py#LL16C11-L16C11) to use `boto3`. One challenge, though, is that `boto3` library is not async (and I wonder whether that was a contributing factor initially for deciding to manually craft the AWS request). However, this can be easily addressed by employing `kombu`'s event loop. Below is my re-implementation of the [`receive_message` method](https://github.com/celery/kombu/blob/main/kombu/asynchronous/aws/sqs/connection.py#L59-L77) employing it:
> 
> ```python
>     def receive_message(
>         self, queue_url, number_messages=1, visibility_timeout=None,
>         attributes=('ApproximateReceiveCount',), wait_time_seconds=None,
>         callback=None
>     ):
>         def make_request(
>             queue_url, number_messages, wait_time_seconds, visibility_timeout,
>             attributes, callback
>         ):
>             kwargs = {
>                 "QueueUrl": queue_url,
>                 "MaxNumberOfMessages": number_messages,
>                 "MessageAttributeNames": attributes,
>                 "WaitTimeSeconds": wait_time_seconds,
>             }
>             if visibility_timeout:
>                 kwargs["VisibilityTimeout"] = visibility_timeout
>             resp = self.sqs_connection.receive_message(**kwargs)
>             if callback:
>                 callback(resp)
> 
>         # Let kombu's event loop call `make_request` asynchronously.
>         return self.hub.call_soon(
>             make_request, 
>             queue_url, number_messages, wait_time_seconds, visibility_timeout,
>             attributes, callback,
>         )
> ```
> 
> I tested this and confirmed it is fixing the issue. However, one thing that still needs to be tested is with regard to multi-threading and multi-processing. `kombu`'s SQS broker [uses a boto3 session](https://github.com/celery/kombu/blob/main/kombu/transport/SQS.py#L674-L679) to communicate with SQS. Unfortunately, boto3 sessions (and resources) are [not thread safe](https://boto3.amazonaws.com/v1/documentation/api/1.19.0/guide/clients.html#multithreading-or-multiprocessing-with-clients); only clients are. Furthermore, when it comes to multi-processing all of them are not safe to be shared.
> 
> Based on my testing, the same process and thread execute both the `call_soon` and `make_request` methods, which makes sense since the SQS broker [relies on the event loop](https://github.com/celery/kombu/blob/main/kombu/transport/SQS.py#L555) to communicate with SQS, and my modified version of `receive_message` also relies on the same event loop. However:
> 
> 1. Is it possible that the same event loop might end up employing more than one thread or process for some reason? I quickly checked the implementation in [hub.py](https://github.com/celery/kombu/blob/main/kombu/asynchronous/hub.py) and that doesn't seem to be the case.
> 2. Is it possible to have more than one event loop? That doesn't seem to be the case either, since the [`set_event_loop` method](https://github.com/celery/kombu/blob/main/kombu/asynchronous/hub.py#L49-L53) uses a global variable, but I cannot be fully confident considering my limited understanding of Celery code base.
> 3. Even if the answer to both questions above is "No", I am not very comfortable about `SQS.py` making the assumption that the event loop is going to be unique, as this is an implicit assumption and, worse, it is very hard to discover without in depth understanding of how async IO is implemented within `kombu`. So, it is theoretically possible that in the future multiple event loops are introduced (assuming that is not the case already) to achieve more IO concurrency.
> 
> Looking forward to hearing from you soon. I am happy to discuss this with you over a meeting to move things faster.

I think you have done a great job here. in future we will swittch to async io based approach. I think we can start with what you have discussed so far. kombu/celery was designed back in 2009 when there was no standard async. the async features here are mostly monkey patched versions which are not so reliable. so we would better off to something more mainstream & standard. if you can come with a draft PR, it will be easier for me to analyze it further alongside you.

**Comment by auvipy** [comment]
Also this could be a great alternative for future https://github.com/terrycain/aioboto3

**Comment by rafidka** [comment]
@auvipy 

> I think you have done a great job here. in future we will swittch to async io based approach. I think we can start with what you have discussed so far. kombu/celery was designed back in 2009 when there was no standard async. the async features here are mostly monkey patched versions which are not so reliable. so we would better off to something more mainstream & standard. if you can come with a draft PR, it will be easier for me to analyze it further alongside you.

Thanks for the confirmation. I will publish a PR for this soon.

One question I would like to ask is with regards to clean up. I see a lot of functions unnecessary functions in the SQS async client. I am willing to delete them, as opposed to spending time re-implementing them using boto3. I confirmed they are not being used within Celery, but can you confirm external dependencies are not supposed to access those functions directly? (There is still the concern some libraries might still use them regardless, e.g. what happens with kombu accessing internal functions with boto3, but I believe I still believe it is better to clean up the code.)

**Comment by rafidka** [comment]
@auvipy , there you go https://github.com/celery/kombu/pull/1759. Sorry for the delay.

**Comment by auvipy** [comment]
it is always better to use public API. also it will be great to get help from AWS folks to improve SQS support in kombu & celery. currently we lack on that part.

**Comment by rafidka** [comment]
> it is always better to use public API. also it will be great to get help from AWS folks to improve SQS support in kombu & celery. currently we lack on that part.

I will need to discuss this my manager before making any commitment. Do you have anything spec you want to get improved, or just in general?

**Comment by auvipy** [comment]
> Do you have anything spec you want to get improved, or just in general?

if you guys could guide us regarding the best practices for the SQS and related parts in kombu & celery, it would be great.
